x-pack/plugins/alerting/server/rules_client/methods/bulk_edit.ts

```diff
@@ -302,209 +306,419 @@ async function bulkEditOcc<Params extends RuleTypeParams>(
     );
 
   const rules: Array<SavedObjectsBulkUpdateObject<RawRule>> = [];
+  const skipped: BulkActionSkipResult[] = [];
   const errors: BulkOperationError[] = [];
-  const apiKeysToInvalidate: string[] = [];
-  const apiKeysMap = new Map<string, { oldApiKey?: string; newApiKey?: string }>();
+  const apiKeysMap: ApiKeysMap = new Map();
   const username = await context.getUserName();
 
   for await (const response of rulesFinder.find()) {
     await pMap(
       response.saved_objects,
-      async (rule) => {
-        try {
-          if (rule.attributes.apiKey) {
-            apiKeysMap.set(rule.id, { oldApiKey: rule.attributes.apiKey });
-          }
+      async (rule: SavedObjectsFindResult<RawRule>) =>
+        updateRuleAttributesAndParamsInMemory({
+          context,
+          rule,
+          operations,
+          paramsModifier,
+          apiKeysMap,
+          rules,
+          skipped,
+          errors,
+          username,
+        }),
+      { concurrency: API_KEY_GENERATE_CONCURRENCY }
+    );
+  }
 
-          const ruleType = context.ruleTypeRegistry.get(rule.attributes.alertTypeId);
+  const { result, apiKeysToInvalidate } =
+    rules.length > 0
+      ? await saveBulkUpdatedRules(context, rules, apiKeysMap)
+      : {
+          result: { saved_objects: [] },
+          apiKeysToInvalidate: [],
+        };
+
+  return {
+    apiKeysToInvalidate,
+    resultSavedObjects: result.saved_objects,
+    errors,
+    rules,
```

Comment:
Do we need both `rules` and `resultSavedObjects` fields here?
It's hard to tell if we use it from our side. I did not find evidence still.
Maybe you use both?

Reply:
Yes, we need both in the `retryIfBulkEditConflicts` function in `x-pack/plugins/alerting/server/rules_client/common/retry_if_bulk_edit_conflicts.ts`